src/main/java/com/google/sps/database/DatabaseImpl.java

```diff
@@ -0,0 +1,190 @@
+/*
+ * Implementation for DatabaseImpl class
+ *  Current public functions at a glance:
+ *  test()
+ *  addVideo()
+ *  addClause()
+ *  addClauses()
+ *  addMetadata()
+ *  getAllKeywords()
+ *  getTimesForKeyword()
+ * 
+ */
+
+package com.google.sps.database;
+
+import com.google.appengine.api.datastore.*;
+import java.util.*;
+
+public class DatabaseImpl implements DatabaseInterface {
+
+    
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    //================================================================================
+    // General Test Function (not used for official testing)
+    //================================================================================
+    
+    public void test() {
+        
+        Map<String, List<Long>> hm = new HashMap<String, List<Long>>();
+
+        String caption = "John Cena";
+        List<Long> timestamps = new ArrayList<>();
+        timestamps.add(42L);
+        timestamps.add(365L);
+        timestamps.add(31415L);
+        hm.put(caption, timestamps);
+
+        String caption2 = "Dwayne 'the Rock' Johnson";
+        List<Long> timestamps2 = new ArrayList<>();
+        timestamps2.add(0L);
+        timestamps2.add(123L);
+        timestamps2.add(404L);
+        hm.put(caption2, timestamps2);
+
+        Entity myEnt = new Entity("video","ncbb5B85sd2");
+
+        datastore.put(myEnt);
+
+        List<Entity> le = new ArrayList<>();
+
+        for (Map.Entry<String, List<Long>> me : hm.entrySet()) {
+            Entity tempEnt = new Entity("caption", me.getKey(), myEnt.getKey());
+            tempEnt.setProperty("timestamps", me.getValue());
+            datastore.put(tempEnt);
+            le.add(tempEnt);
+        }
+        
+        for (Entity e : le) {
+            System.out.println("Parent: " + e.getParent());
+        }
+
+    }
+
+    //================================================================================
+    // Public Interface Functions (see DatabaseInterface.java for official descriptions)
+    //================================================================================    
+
+    // add a video to the database
+    public int addVideo(String videoID, String metadata) { 
+        Entity vidEnt = new Entity("video", videoID); //error possibility
+        datastore.put(vidEnt);  //error possibility
+
+        Entity metaEnt = new Entity("metadata", metadata, vidEnt.getKey()); //error possibility
+        datastore.put(metaEnt); //error possibility
+
+        return 0;
+    }
+
+    // add a keyphrase + timestamp pair to a particular video in the db
+    public int addClause(String videoID, String keyword, List<Long> timestamps) {
+        Entity vidEnt = getVideo(videoID);  //error possibility
+        Entity keyEnt = new Entity("caption", keyword, vidEnt.getKey()); //error possibility
```

Comment:
Same for all other times in the file where an entity kind is a hard-coded string